tests/test_bolt2-01-close_channel.py

```diff
@@ -0,0 +1,187 @@
+                # why the script is different
+                ExpectMsg("shutdown", channel_id=channel_idx),
+            ],
+        ),
+    ]
+    run_runner(runner, merge_events_sequences(pre=pre_events, post=test))
+
+
+def test_close_channel_shutdown_msg_wrong_script_pubkey_received_side(
+    runner: Runner, namespaceoverride: Any
+) -> None:
+    """Test close operation from the receiver view point, in the case when
+    the sender set a wrong script pub key not specified in the spec.
+     ______________________________________________________
+    | runner -> shutdown (wrong script pub key) -> ln-node |
+    | runner <-         warning msg             <- ln-node |
+    -------------------------------------------------------
+    """
+    namespaceoverride(peer_message_namespace())
+
+    # the option that the helper method feels for us
+    test_opts = {}
+    pre_events_conn = connect_to_node_helper(
+        runner, tx_spendable=tx_spendable, conn_privkey="03"
+    )
+    pre_events = open_and_announce_channel_helper(
+        runner, conn_privkey="03", opts=test_opts
+    )
+    # merge the two events
+    pre_events = merge_events_sequences(pre_events_conn, pre_events)
+    channel_idx = channel_id()
+
+    script = BitcoinUtils.build_valid_script(ScriptType.INVALID_CLOSE_SCRIPT)
+
+    test = [
+        # runner sent shutdown message to the ln implementation
+        TryAll(
+            [
+                Msg(
+                    "shutdown",
+                    channel_id=channel_idx,
+                    scriptpubkey=script,
+                ),
+                MustNotMsg("shutdown"),
+                MustNotMsg("add_htlc"),
+                # FIXME: add support for warning messages in pyln package?
+            ],
+        ),
+    ]
+    run_runner(runner, merge_events_sequences(pre=pre_events, post=test))
+
+
+def test_close_channel_allow_multiple_shutdown_msg_receive_side(
+    runner: Runner, namespaceoverride: Any
+) -> None:
+    """
+    Close operation from the receiver point of view, where the receiver will
+    receive multiple shutdown msg.
+
+    FIXME: this need to be allowed by the spec!
+     ________________________________
+    | runner -> shutdown -> ln-node |
+    | runner -> shutdown -> ln-node |
+    | runner <- shutdown <- ln-node |
+    --------------------------------
+    """
+    namespaceoverride(peer_message_namespace())
+    test_opts = {}
+    pre_events_conn = connect_to_node_helper(
+        runner, tx_spendable=tx_spendable, conn_privkey="03"
+    )
+    pre_events = open_and_announce_channel_helper(
+        runner, conn_privkey="03", opts=test_opts
+    )
+    # merge the two events
+    pre_events = merge_events_sequences(pre_events_conn, pre_events)
+    channel_idx = channel_id()
+
+    script = BitcoinUtils.build_valid_script()
+
+    test = [
+        # runner sent shutdown message to the ln implementation
+        TryAll(
+            [
+                Msg(
```

Comment:
Hi I dont understand why there is a need for duplicate msg shutdown?

Reply:
We are testing the following use case

```
     ________________________________
    | runner -> shutdown -> ln-node |
    | runner -> shutdown -> ln-node |
    | runner <- shutdown <- ln-node |
    --------------------------------
```

The spec doesn't tell us anything about multiple shutdowns, but this can be common if the other peers are slow to answer.

I'm writing also a small proposal for the spec about this use case